### PR TITLE
Env: Add support for running in themes.

### PR DESCRIPTION
**Pull request by epiqueras**
Follows #17724

## Description

This PR adds support for running `@wordpress/env` in theme repositories.

It does this by inferring what it's running on by looking for plugin or theme headers in the current working directory.

## How has this been tested?

It was verified that `@wordpress/env` can correctly infer whether it is running in a theme or plugin and configure itself accordingly.

## Types of Changes

*New Feature:* `@wordpress/env` Now supports theme repositories.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

File: packages/env/lib/detect-context.js [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+const finished = util.promisify( stream.finished );
+
+module.exports = async function detectContext() {
+	const context = {};
+
+	// Race multiple file read streams against each other until
+	// a plugin or theme header is found.
+	const files = ( await readDir( './' ) ).filter( ( file ) => {
+		const extension = path.extname( file );
+		return extension === '.php' || extension === '.css';
```
Comment (swissspidy) [inline packages/env/lib/detect-context.js]: One small optimization for themes would be to look for `style.css` specifically, not just any CSS file.

Comment (epiqueras) [reply packages/env/lib/detect-context.js]: Thanks for taking a look. I changed it here: https://github.com/WordPress/gutenberg/pull/17732/commits/a7f97af35719061ef3656eb7356936b9571fadcd.

File: packages/env/lib/detect-context.js [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+			}
+		} );
+		streams.push( fileStream );
+	}
+	await Promise.all(
+		streams.map( ( fileStream ) => finished( fileStream ).catch( () => {} ) )
+	);
+
+	return context;
+};
```
[the remark this answers is not in the document]
Comment (epiqueras) [reply packages/env/lib/detect-context.js]: >then the logic is to check for the plugin's main .php file based on the directory name which matches the slug.

The directory name doesn't always match the slug.

>which traverses up the directory tree to find the package root dir since the running script dir doesn't always match the current working dir.

You're supposed to run the tool in the plugin you want to develop so that wouldn't really make sense in this context.

>they limit the data of the file to 8kb on read.

This could make sense if we run into performance issues.

>lower system especially for large amounts of files

We only check root `.php` files and `style.css` so I doubt this will ever be an issue.

**Comment by epiqueras** [comment]
Thanks!

File: packages/env/package.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"name": "@wordpress/env",
-	"version": "1.0.0",
+	"version": "0.0.0",
```
Comment (ellatrix) [inline packages/env/package.json]: Is this change intentional?

Comment (epiqueras) [reply packages/env/package.json]: Yes, so that the release tool can bump it for the first release.

**Review by ellatrix — APPROVED** [review]
Nice improvement!
